internal/command/user_human_password.go

```diff
@@ -100,38 +104,36 @@ func (c *Commands) ChangePassword(ctx context.Context, orgID, userID, oldPasswor
 	if err != nil {
 		return nil, err
 	}
-	if wm.EncodedHash == "" {
-		return nil, caos_errs.ThrowPreconditionFailed(nil, "COMMAND-Fds3s", "Errors.User.Password.Empty")
-	}
-	if err = c.canUpdatePassword(ctx, newPassword, wm); err != nil {
-		return nil, err
-	}
 
-	ctx, spanPasswap := tracing.NewNamedSpan(ctx, "passwap.VerifyAndUpdate")
-	updated, err := c.userPasswordHasher.VerifyAndUpdate(wm.EncodedHash, oldPassword, newPassword)
-	spanPasswap.EndWithError(err)
-	if err = convertPasswapErr(err); err != nil {
+	// if updated hash ignore, as the new password is set afterward
+	if _, err = c.verifyPassword(ctx, wm.EncodedHash, oldPassword); err != nil {
 		return nil, err
 	}
-	err = c.pushAppendAndReduce(ctx, wm,
-		user.NewHumanPasswordChangedEvent(ctx, UserAggregateFromWriteModel(&wm.WriteModel), updated, false, userAgentID))
-	if err != nil {
-		return nil, err
+	return c.setPassword(ctx, wm, newPassword, false)
```

Comment:
VerifyAndUpdate should remain in use here, as you risk wasting time and resources on 1 additional hash computation.

For example consider the following events:
1. Zitadel changes hash config, cost etc.
2. User changes password
3. Old password is verified (Hash operation 1)
4. Old password is re-hashed for config updated (Hash operation 2)
5. New password is hashed with random salt (Hash operation 3)

Instead, when VerifyAndUpdate passwap knows not to execute step 4 and save 1 hash operation by returning a hash of the new password directly.

Reply:
Used verifyAndUpdate again, also changed it in user_v2_human.go